subsys/net/lib/sockets/sockets_raw.c

```diff
@@ -0,0 +1,351 @@
+		errno = EAGAIN;
+		return -1;
+	}
+
+	/* Register the callback before sending in order to receive the response
+	 * from the peer.
+	 */
+	err = net_context_recv(ctx, zraw_received_cb, K_NO_WAIT,
+			       ctx->user_data);
+	if (err < 0) {
+		net_pkt_unref(send_pkt);
+		errno = -err;
+		return -1;
+	}
+
+	err = net_context_sendto(send_pkt, dest_addr, addrlen, NULL,
+				 timeout, NULL, ctx->user_data);
+
+	if (err < 0) {
+		net_pkt_unref(send_pkt);
+		errno = -err;
+		return -1;
+	}
+
+	return len;
+}
+
+ssize_t zraw_recvfrom_ctx(struct net_context *ctx, void *buf, size_t max_len,
+			  int flags, struct sockaddr *src_addr,
+			  socklen_t *addrlen)
+{
+	size_t recv_len = 0;
+	s32_t timeout = K_FOREVER;
+	struct net_pkt *pkt;
+
+	if ((flags & ZSOCK_MSG_DONTWAIT) || sock_is_nonblock(ctx)) {
+		timeout = K_NO_WAIT;
+	}
+
+	if (flags & ZSOCK_MSG_PEEK) {
+		int res;
+
+		res = _k_fifo_wait_non_empty(&ctx->recv_q, timeout);
+		/* EAGAIN when timeout expired, EINTR when cancelled */
+		if (res && res != -EAGAIN && res != -EINTR) {
+			errno = -res;
+			return -1;
+		}
+
+		pkt = k_fifo_peek_head(&ctx->recv_q);
+	} else {
+		pkt = k_fifo_get(&ctx->recv_q, timeout);
+	}
+
+	if (!pkt) {
+		errno = EAGAIN;
+		return -1;
+	}
+
+	/* We do not handle any headers here,
+	 * just pass the whole packet to caller.
+	 */
+	recv_len = net_pkt_get_len(pkt);
+	if (recv_len > max_len) {
+		recv_len = max_len;
+	}
+
+	net_frag_linearize(buf, recv_len, pkt, 0, recv_len);
+
+	if (!(flags & ZSOCK_MSG_PEEK)) {
+		net_pkt_unref(pkt);
+	}
+
+	return recv_len;
+}
+
+int zraw_getsockopt_ctx(struct net_context *ctx, int level, int optname,
+			void *optval, socklen_t *optlen)
+{
+	if (!optval || !optlen) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	return sock_fd_op_vtable.getsockopt(ctx, level, optname,
+					    optval, optlen);
+}
+
+int zraw_setsockopt_ctx(struct net_context *ctx, int level, int optname,
+			const void *optval, socklen_t optlen)
+{
+	return sock_fd_op_vtable.setsockopt(ctx, level, optname,
+					    optval, optlen);
+}
+
+static ssize_t raw_sock_read_vmeth(void *obj, void *buffer, size_t count)
+{
+	return zraw_recvfrom_ctx(obj, buffer, count, 0, NULL, 0);
+}
+
+static ssize_t raw_sock_write_vmeth(void *obj, const void *buffer,
+				    size_t count)
+{
+	return zraw_sendto_ctx(obj, buffer, count, 0, NULL, 0);
+}
+
+static int raw_sock_ioctl_vmeth(void *obj, unsigned int request, va_list args)
+{
+	return sock_fd_op_vtable.fd_vtable.ioctl(obj, request, args);
+}
+
+/*
+ * TODO: A raw socket can be bound to a network device using SO_BINDTODEVICE.
+ */
+static int raw_sock_bind_vmeth(void *obj, const struct sockaddr *addr,
+			       socklen_t addrlen)
+{
+	return zraw_bind_ctx(obj, addr, addrlen);
+}
+
+/* The connect() function is no longer necessary. */
+static int raw_sock_connect_vmeth(void *obj, const struct sockaddr *addr,
+				  socklen_t addrlen)
+{
+	return 0;
+}
+
+/*
+ * The listen() and accept() functions are without any functionality,
```

Comment:
If they're "without any functionality", why did you implement them on net_context level?

Reply:
Wasn't addressed.